### DISCUSSION TITLE
Is any way to convert .mdx to .html without nextjs?

### DISCUSSION THREAD

**Discussion by zfben** [mdx-js/mdx#1985] [Q&A] [answered]
I found most of mdx packages are convert .mdx to js code, how should I convert it to static HTML for SEO? (And if I don't use Next.js)

Thank you.

**Comment by wooorm** [comment, marked as answer]
There are many ways. Which way works for you, depends on what you do use.

**Reply by zfben** [reply]
Is there a pure way? I'm working on a tool only depends on mdx, doesn't depends on other tools.

**Reply by wooorm** [reply]
Yes. There are several “pure” ways. Please explain what you are making.

**Reply by zfben** [reply]
My parser's code is:

```ts
import { compileSync } from '@mdx-js/mdx'
import remarkGfm from 'remark-gfm'
import rehypeHighlight from 'rehype-highlight'

export function parse (body: string): string {
  return String(compileSync(body, {
    format: 'mdx',
    outputFormat: 'function-body',
    useDynamicImport: true,
    remarkPlugins: [remarkGfm],
    rehypePlugins: [rehypeHighlight]
  }))
}
```

then the function-body will be displayed in a react app.

Now I want to convert the mdx content to HTML code with the above plugins.

**Reply by wooorm** [reply]
I believe you should read our quick start in the docs: https://mdxjs.com/docs/getting-started/#quick-start

**Reply by zfben** [reply]
I see the Site generators section in the document, but they use 3rd packages.
Can I generate static HTML without these 3rd tools?

**Reply by wooorm** [reply]
Please properly read the “Quick start” section. In particular, see the list following “You can also use MDX if you’re not using a bundler:” Then, see the following subsection too: “JSX”.

To summarize: you can import MDX in Node or use `evaluate` from `mdx-js/mdx`. You have to choose a JSX runtime though. React, Preact, Svelte, Vue, etcetera.

**Reply by zfben** [reply]
My problem is `mdx-js/mdx` generate javascript code, not html code. I want to generate HTML code for better SEO.

**Reply by wooorm** [reply]
No it doesn’t. importing in Node and `evaluate` **do not return code**. They compile and evaluate code. **they return components**. You can render those components to get HTML.

**Reply by zfben** [reply]
@wooorm Thanks a lot. Finally, I know how to generate HTML.

```ts
import remarkGfm from 'remark-gfm'
import rehypeHighlight from 'rehype-highlight'
import { renderToString } from 'react-dom/server'
import * as runtime from 'react/jsx-runtime.js'
import { evaluateSync } from '@mdx-js/mdx'
import { createElement } from 'react'

export function generate (body: string) {
  const mdx = evaluateSync(body, {
    ...runtime as any,
    remarkPlugins: [remarkGfm],
    rehypePlugins: [rehypeHighlight]
  }).default

  return renderToString(createElement(mdx))
}

```

**Reply by ManasMakde** [reply]
I tried @zfben 's [code](https://github.com/orgs/mdx-js/discussions/1985#discussioncomment-2437016) but it gave me the following error:

> Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './jsx-runtime.js' is not defined by "exports"

To fix it change it to `import * as runtime from 'react/jsx-runtime'`

**Reply by flandia** [reply]
Use `renderToStaticMarkup` instead of `renderToString` to prevent React from adding React internal attributes.
